plugins/shared/loader/loader.go

```diff
@@ -0,0 +1,261 @@
+type PluginInstance struct {
+	client   *plugin.Client
+	instance interface{}
+}
+
+// Internal returns if the plugin is internal
+func (p *PluginInstance) Internal() bool {
+	return p.client == nil
+}
+
+// Kill kills the plugin if it is external. It is safe to call on internal
+// plugins.
+func (p *PluginInstance) Kill() {
+	if p.client != nil {
+		p.client.Kill()
+	}
+}
+
+// ReattachConfig returns the ReattachConfig and whether the plugin is internal
+// or not. If the second return value is true, no ReattachConfig is possible to
+// return.
+func (p *PluginInstance) ReattachConfig() (*plugin.ReattachConfig, bool) {
+	if p.client == nil {
+		return nil, false
+	}
+	return p.client.ReattachConfig(), true
+}
+
+// Plugin returns the wrapped plugin instance.
+func (p *PluginInstance) Plugin() interface{} {
+	return p.instance
+}
+
+// PluginLoader is used to retrieve plugins either externally or from internal
+// factories.
+type PluginLoader struct {
+	// logger is the plugin loaders logger
+	logger log.Logger
+
+	// pluginDir is the directory containing plugin binaries
+	pluginDir string
+
+	// plugins maps a plugin to information required to launch it
+	plugins map[PluginID]*pluginInfo
+}
+
+// PluginID is a tuple identifying a plugin
+type PluginID struct {
+	// Name is the name of the plugin
+	Name string
+
+	// PluginType is the plugin's type
+	PluginType string
+}
+
+// String returns a friendly representation of the plugin.
+func (id PluginID) String() string {
+	return fmt.Sprintf("%q (%v)", id.Name, id.PluginType)
+}
+
+// PluginLoaderConfig configures a plugin loader.
+type PluginLoaderConfig struct {
+	// Logger is the logger used by the plugin loader
+	Logger log.Logger
+
+	// PluginDir is the directory scanned for loading plugins
+	PluginDir string
+
+	// Configs is an optional set of configs for plugins
+	Configs []*config.PluginConfig
+
+	// InternalPlugins allows registering internal plugins.
+	InternalPlugins map[PluginID]*InternalPluginConfig
+}
+
+// InternalPluginConfig is used to configure launching an internal plugin.
+type InternalPluginConfig struct {
+	Config  map[string]interface{}
+	Factory plugins.PluginFactory
+}
+
+// pluginInfo captures the necessary information to launch and configure a
+// plugin.
+type pluginInfo struct {
+	factory plugins.PluginFactory
+
+	exePath string
+	args    []string
+
+	baseInfo *base.PluginInfoResponse
+	version  *version.Version
+
+	configSchema  *hclspec.Spec
+	config        map[string]interface{}
+	msgpackConfig []byte
+}
+
+// NewPluginLoader returns an instance of a plugin loader or an error if the
+// plugins could not be loaded
+func NewPluginLoader(config *PluginLoaderConfig) (*PluginLoader, error) {
+	if err := validateConfig(config); err != nil {
+		return nil, fmt.Errorf("invalid plugin loader configuration passed: %v", err)
+	}
+
+	logger := config.Logger.Named("plugin-loader").With("plugin-dir", config.PluginDir)
```

Comment:
Hm, I've been using underscores as a separator in log names/keys so it looks like an identifier and matches hcl params (when applicable): https://gist.github.com/schmichael/da63914230b5127fb26486b01446f5a5

I don't mind switching, but we should probably standardize.